File: matiasmasso/Mat.NET/DT2/DTOAddress.vb
Public Class DTOAddress
    Property viaNom As String
    Property num As String
    Property pis As String

    Property text As String
    Property zip As DTOZip
    Property coordenadas As GeoHelper.Coordenadas

    Property src As DTOBaseGuid
    Property codi As Codis

    Property isNew As Boolean

    Public Enum Codis
        NotSet
        Fiscal
        Correspondencia
        Entregas
    End Enum

    Shared Function Factory(oSrc As DTOBaseGuid, oCodi As DTOAddress.Codis) As DTOAddress
        Dim retval As New DTOAddress
        With retval
            .Src = oSrc
            .Codi = oCodi
        End With
        Return retval
    End Function

    Public Shadows Function Equals(oCandidate As Object) As Boolean
        Dim retval As Boolean
        If oCandidate IsNot Nothing Then
            If TypeOf oCandidate Is DTOAddress Then
                Dim oAddress As DTOAddress = oCandidate
                If _Text = oAddress.Text Then
                    Dim oZip As DTOZip = oAddress.Zip
                    retval = _Zip.Equals(oZip)
                End If
            End If
        End If
        Return retval
    End Function


    Public Function ExportCod() As DTOInvoice.ExportCods
        Dim retval As DTOInvoice.ExportCods = DTOInvoice.ExportCods.NotSet

        If _Zip IsNot Nothing Then
            Dim oLocation As DTOLocation = _Zip.Location
            If oLocation IsNot Nothing Then
                Dim oZona As DTOZona = oLocation.Zona
                If oZona IsNot Nothing Then
                    retval = oZona.ExportCod
                End If
            End If
        End If
        Return retval
    End Function

    Shared Function ExportCod(oAddress As DTOAddress) As DTOInvoice.ExportCods
        Dim retval As DTOInvoice.ExportCods = DTOInvoice.ExportCods.NotSet
        If oAddress IsNot Nothing Then
            retval = oAddress.ExportCod()
        End If
        Return retval
    End Function

    Shared Function FullText(oAddress As DTOAddress, oLang As DTOLang) As String
        Dim retval As String = oAddress.toMultilineString(oLang).Replace(vbCrLf, " - ")
        Return retval
    End Function

    Public Function toMultilineString(oLang As DTOLang) As String
        Dim sb As New Text.StringBuilder
        sb.AppendLine(_Text)
        sb.AppendLine(_Zip.FullNom(oLang))
        Dim retval = sb.ToString
        Return retval
    End Function

    Shared Function TextOrDefault(oAddress As DTOAddress) As String
        Dim retval As String = ""
        If oAddress IsNot Nothing Then
            retval = oAddress.Text
        End If
        Return retval
    End Function

    Shared Function ZipyCit(oAddress As DTOAddress) As String
        Dim retval As String = ""
        If oAddress IsNot Nothing AndAlso oAddress.Zip IsNot Nothing Then
            retval = oAddress.Zip.ZipyCit
        End If
        Return retval
    End Function

    Public Function SingleLineText() As String
        'Dim lines As String() = _text.Split(vbCrLf)
        Dim retval = _text.Replace(vbCrLf, " ").Replace(vbLf, " ")
        Return retval
    End Function

    Public Function Location() As DTOLocation
        Dim retval As DTOLocation = Nothing
        If Zip IsNot Nothing Then
            retval = _Zip.Location
        End If
        Return retval
    End Function

    Shared Function Location(oAddress As DTOAddress) As DTOLocation
        Dim retval As DTOLocation = Nothing
        If oAddress IsNot Nothing Then
            retval = oAddress.Location
        End If
        Return retval
    End Function

    Public Function Zona() As DTOZona
        Dim retval As DTOZona = Nothing
        Dim oLocation = Me.Location
        If oLocation IsNot Nothing Then
            retval = oLocation.Zona
        End If
        Return retval
    End Function

    Shared Function Zona(oAddress As DTOAddress) As DTOZona
        Dim retval As DTOZona = Nothing
        If oAddress IsNot Nothing Then
            retval = oAddress.Zona
        End If
        Return retval
    End Function

    Public Function Provincia() As DTOAreaProvincia
        Dim retval As DTOAreaProvincia = Nothing
        Dim oZona = Me.Zona
        If oZona IsNot Nothing Then
            retval = oZona.Provincia
        End If
        Return retval
    End Function

    Shared Function Provincia(oAddress As DTOAddress) As DTOAreaProvincia
        Dim retval As DTOAreaProvincia = Nothing
        If oAddress IsNot Nothing Then
            retval = oAddress.Provincia
        End If
        Return retval
    End Function

    Public Function Country() As DTOCountry
        Dim retval As DTOCountry = Nothing
        Dim oZona = Me.Zona
        If oZona IsNot Nothing Then
            retval = oZona.Country
        End If
        Return retval
    End Function

    Shared Function Country(oAddress As DTOAddress) As DTOCountry
        Dim retval As DTOCountry = Nothing
        If oAddress IsNot Nothing Then retval = oAddress.Country
        Return retval
    End Function

    Public Function countryNom(oLang As DTOLang) As String
        Dim retval As String = ""
        If _zip IsNot Nothing Then
            retval = _zip.CountryNom(oLang)
        End If
        Return retval
    End Function

    Public Function suggestedLang() As DTOLang
        Dim retval As DTOLang = Nothing
        Dim oZona As DTOZona = Me.Zona
        If oZona Is Nothing Then
            retval = DTOLang.ESP
        Else
            retval = oZona.suggestedLang
        End If
        Return retval
    End Function

    Shared Function LocationFullNom(oAddress As DTOAddress, oLang As DTOLang) As String
        Dim retval As String = ""
        If oAddress IsNot Nothing AndAlso oAddress.Zip IsNot Nothing AndAlso oAddress.Zip.Location IsNot Nothing Then
            retval = DTOLocation.FullNom(oAddress.Zip.Location, oLang)
        End If
        Return retval
    End Function


    Shared Function ClxLocation(oAddress As DTOAddress) As String
        'per generar Contact.Fullnom
        Dim sb As New Text.StringBuilder
        If oAddress IsNot Nothing Then
            If oAddress.Zip IsNot Nothing Then
                Dim oLocation = oAddress.Zip.Location
                If oLocation IsNot Nothing Then
                    Dim sCit As String = oLocation.Nom
                    sb.Append(sCit)
                    Dim oZona = oLocation.Zona
                    If oZona IsNot Nothing Then
                        Dim oCountry = oZona.Country
                        If oCountry Is Nothing Then
                        ElseIf oCountry.Equals(DTOCountry.wellknown(DTOCountry.wellknowns.Spain)) Then
                            If oZona.Provincia IsNot Nothing Then
                                Dim sProvincia As String = oZona.Provincia.Nom
                                If sProvincia <> "" And sProvincia <> oLocation.Nom Then
                                    sb.Append(String.Format(", {0}", sProvincia))
                                End If
                            End If
                        Else
                            Dim sCountry As String = oCountry.langNom.Esp
                            If sCountry > "" Then
                                sb.Append(String.Format(", {0}", sCountry))
                            End If
                        End If
                    End If
                End If
            End If
        End If

        Dim retval = sb.ToString
        Return retval
    End Function

    Public Function IsEmpty() As Boolean
        Dim retval As Boolean = _Text = "" And _ViaNom = "" And _Num = "" And _Zip Is Nothing
        Return retval
    End Function

    Shared Function Lines(oContact As DTOContact) As List(Of String)
        Dim retval As New List(Of String)
        With retval
            .Add(IIf(oContact.Nom = "", oContact.NomComercial, oContact.Nom))
            .Add(oContact.Address.Text)
            .Add(oContact.Address.Zip.ZipyCit())
        End With
        Return retval
    End Function

    Shared Function ProvinciaOPais(oZip As DTOZip) As String
        Dim retval As String = ""
        If oZip IsNot Nothing Then
            If DTOArea.IsEsp(oZip) Then
                Dim exs As New List(Of Exception)
                Dim oProvincia = DTOZip.Provincia(oZip)
                If oProvincia IsNot Nothing Then
                    retval = "(" & oProvincia.Nom & ")"
                End If
            Else
                retval = "(" & DTOArea.Country(oZip).langNom.Esp & ")"
            End If
        End If
        Return retval
    End Function

    Public Function IsEsp() As Boolean
        Dim retval As Boolean = DTOArea.IsEsp(_Zip)
        Return retval
    End Function

    Shared Function GoogleText(oAddress As DTOAddress) As String
        Dim sb As New System.Text.StringBuilder
        If oAddress.ViaNom > "" Then
            sb.Append(oAddress.ViaNom)
            sb.Append(",")
        End If
        If oAddress.Num > "" Then
            sb.Append(oAddress.Num)
            sb.Append(",")
        End If
        If oAddress.Zip IsNot Nothing Then
            sb.Append(oAddress.Zip.ZipCod)
            sb.Append(",")
            sb.Append(oAddress.Location.Nom)
            sb.Append(",")
            sb.Append(oAddress.CountryNom(DTOLang.ESP))
        End If
        Dim retval = sb.ToString
        Return retval
    End Function

    Shared Function GoogleNormalized(oSrc As DTOAddress) As String
        Dim sb As New System.Text.StringBuilder
        If oSrc IsNot Nothing AndAlso Not String.IsNullOrEmpty(oSrc.Text) Then
            Dim sAdrSegments() As String = oSrc.Text.Split(" ")
            For Each segment As String In sAdrSegments
                'elimina el segon numero del carrer en cas de varios numeros
                If IsNumeric(segment) Then
                    If segment.Contains("-") Then
                        segment = segment.Substring(0, segment.IndexOf("-"))
                    End If
                End If
            Next
            Dim sAdr As String = String.Join("+", sAdrSegments)
            Dim sCit As String = oSrc.Zip.Location.Nom.Replace(" ", "+")
            Dim sZona As String = oSrc.Zip.Location.Zona.Nom.Replace(" ", "+")
            Dim sCountry As String = oSrc.zip.location.zona.country.langNom.Esp.Replace(" ", "+")
            sb.Append(sAdr)
            If oSrc.Zip.ZipCod > "" Then sb.Append("," & oSrc.Zip.ZipCod)
            sb.Append("," & sCit)
            sb.Append("," & sZona)
            sb.Append("," & sCountry)
        End If
        Dim retval As String = sb.ToString
        Return retval

    End Function

    Shared Function MultiLine(oAddress As DTOAddress) As String
        Dim sb As New System.Text.StringBuilder
        If oAddress IsNot Nothing Then
            If oAddress.Text > "" Then
                sb.AppendLine(oAddress.Text.Trim)
            End If
            sb.Append(DTOZip.FullNom(oAddress.Zip))
        End If
        Dim retval As String = sb.ToString
        Return retval
    End Function

    Public Function ToHtml() As String
        Dim sLines = MultiLine(Me)
        Dim retval As String = TextHelper.ToHtml(sLines)
        Return retval
    End Function


    Shared Function ReverseSingleLine(oAddress As DTOAddress, oLang As DTOLang) As String
        Dim sb As New Text.StringBuilder
        Dim sLocation As String = DTOAddress.LocationFullNom(oAddress, oLang)
        If sLocation > "" Then
            sb.Append(sLocation)
            If oAddress.Text > "" Then
                sb.Append(", ")
            End If
        End If
        sb.Append(oAddress.Text)
        Dim retval As String = sb.ToString
        Return retval
    End Function

    Shared Function FullText(oAddress As DTOAddress) As String
        Dim retval As String = DTOAddress.MultiLine(oAddress).Replace(vbCrLf, " - ")
        Return retval
    End Function

    Shared Function FullHtml(oAddress As DTOAddress) As String
        Dim retval As String = DTOAddress.MultiLine(oAddress).Replace(vbCrLf, "<br/>")
        Return retval
    End Function


    Public Function Trimmed() As DTOAddress
        Dim oType As Type = Me.GetType()
        Dim retval = Activator.CreateInstance(Me.GetType())
        Try
            Dim oProperties = oType.GetProperties()
            For Each oProperty In oProperties
                If GetType(DTOBaseGuid).IsAssignableFrom(oProperty.PropertyType) Then
                    Try
                        Dim oBaseGuid As DTOBaseGuid = oProperty.GetValue(Me)
                        If oBaseGuid IsNot Nothing Then
                            Dim oTrimmedPropertyValue = Activator.CreateInstance(oProperty.PropertyType, oBaseGuid.Guid)
                            oProperty.SetValue(retval, oTrimmedPropertyValue)
                        End If

                    Catch ex As Exception
                        oProperty.SetValue(retval, oProperty.GetValue(Me))
                    End Try
                Else
                    oProperty.SetValue(retval, oProperty.GetValue(Me))
                End If
            Next

        Catch ex As Exception
            Stop
        End Try
        Return retval
    End Function
End Class

File: matiasmasso/Mat.NET/DTO/DTOAddress.cs
using MatHelperStd;
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOAddress
    {
        public string ViaNom { get; set; }
        public string Num { get; set; }
        public string Pis { get; set; }

        public string Text { get; set; }
        public DTOZip Zip { get; set; }
        public GeoHelper.Coordenadas Coordenadas { get; set; }

        public DTOBaseGuid Src { get; set; }
        public Codis Codi { get; set; }

        public bool IsNew { get; set; }

        public enum Codis
        {
            NotSet,
            Fiscal,
            Correspondencia,
            Entregas,
            FraConsumidor
        }

        public static DTOAddress Factory(DTOBaseGuid oSrc, DTOAddress.Codis oCodi)
        {
            DTOAddress retval = new DTOAddress();
            {
                retval.Src = oSrc;
                retval.Codi = oCodi;
            }
            return retval;
        }

        public new bool Equals(object oCandidate)
        {
            bool retval = false;
            if (oCandidate != null)
            {
                if (oCandidate is DTOAddress)
                {
                    DTOAddress oAddress = (DTOAddress)oCandidate;
                    if (Text == oAddress.Text)
                    {
                        DTOZip oZip = oAddress.Zip;
                        retval = Zip.Equals(oZip);
                    }
                }
            }
            return retval;
        }


        public DTOInvoice.ExportCods ExportCod()
        {
            DTOInvoice.ExportCods retval = DTOInvoice.ExportCods.notSet;

            if (Zip != null)
            {
                DTOLocation oLocation = Zip.Location;
                if (oLocation != null)
                {
                    DTOZona oZona = oLocation.Zona;
                    if (oZona != null)
                        retval = oZona.ExportCod;
                }
            }
            return retval;
        }

        public static DTOInvoice.ExportCods ExportCod(DTOAddress oAddress)
        {
            DTOInvoice.ExportCods retval = DTOInvoice.ExportCods.notSet;
            if (oAddress != null)
                retval = oAddress.ExportCod();
            return retval;
        }

        public static string FullText(DTOAddress oAddress, DTOLang oLang)
        {
            return oAddress.Text.toSingleLine();
        }

        public string ToMultilineString(DTOLang oLang)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.AppendLine(Text);
            sb.AppendLine(Zip.FullNom(oLang));
            var retval = sb.ToString();
            return retval;
        }

        public static string TextOrDefault(DTOAddress oAddress)
        {
            string retval = "";
            if (oAddress != null)
                retval = oAddress.Text;
            return retval;
        }


        public static string ZipCod(DTOAddress oAddress)
        {
            string retval = "";
            if (oAddress != null && oAddress.Zip != null)
                retval = oAddress.Zip.ZipCod;
            return retval;
        }

        public static string ZipyCit(DTOAddress oAddress)
        {
            string retval = "";
            if (oAddress != null && oAddress.Zip != null)
                retval = oAddress.Zip.ZipyCit();
            return retval;
        }

        public string SingleLineText()
        {
            string retval = "";
            if (!string.IsNullOrEmpty(Text))
                retval = this.Text.toSingleLine();
            return retval;
        }

        public DTOLocation Location()
        {
            DTOLocation retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (this.Zip != null)
                retval = this.Zip.Location;
            return retval;
        }

        public static DTOLocation Location(DTOAddress oAddress)
        {
            DTOLocation retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (oAddress != null)
                retval = oAddress.Location();
            return retval;
        }

        public DTOZona Zona()
        {
            DTOZona retval = null/* TODO Change to default(_) if this is not a reference type */;
            var oLocation = this.Location();
            if (oLocation != null)
                retval = oLocation.Zona;
            return retval;
        }

        public bool UnEquals(DTOAddress candidate)
        {
            bool retval = !Equals(candidate);
            return retval;
        }

        public static DTOZona Zona(DTOAddress oAddress)
        {
            DTOZona retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (oAddress != null)
                retval = oAddress.Zona();
            return retval;
        }

        public DTOAreaProvincia Provincia()
        {
            DTOAreaProvincia retval = null/* TODO Change to default(_) if this is not a reference type */;
            var oZona = this.Zona();
            if (oZona != null)
                retval = oZona.Provincia;
            return retval;
        }

        public static DTOAreaProvincia Provincia(DTOAddress oAddress)
        {
            DTOAreaProvincia retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (oAddress != null)
                retval = oAddress.Provincia();
            return retval;
        }

        public DTOCountry Country()
        {
            DTOCountry retval = null/* TODO Change to default(_) if this is not a reference type */;
            var oZona = this.Zona();
            if (oZona != null)
                retval = oZona.Country;
            return retval;
        }

        public static DTOCountry Country(DTOAddress oAddress)
        {
            DTOCountry retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (oAddress != null)
                retval = oAddress.Country();
            return retval;
        }

        public string CountryNom(DTOLang oLang)
        {
            string retval = "";
            if (this.Zip != null)
                retval = this.Zip.CountryNom(oLang);
            return retval;
        }

        public DTOLang SuggestedLang()
        {
            DTOLang retval = null/* TODO Change to default(_) if this is not a reference type */;
            DTOZona oZona = this.Zona();
            if (oZona == null)
                retval = DTOLang.ESP();
            else
                retval = oZona.SuggestedLang();
            return retval;
        }

        public static string LocationFullNom(DTOAddress oAddress, DTOLang oLang)
        {
            string retval = "";
            if (oAddress != null && oAddress.Zip != null && oAddress.Zip.Location != null)
                retval = DTOLocation.FullNom(oAddress.Zip.Location, oLang);
            return retval;
        }


        public static string ClxLocation(DTOAddress oAddress)
        {
            // per generar Contact.FullNom
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            if (oAddress != null)
            {
                if (oAddress.Zip != null)
                {
                    var oLocation = oAddress.Zip.Location;
                    if (oLocation != null)
                    {
                        string sCit = oLocation.Nom;
                        sb.Append(sCit);
                        var oZona = oLocation.Zona;
                        if (oZona != null)
                        {
                            var oCountry = oZona.Country;
                            if (oCountry == null)
                            {
                            }
                            else if (oCountry.Equals(DTOCountry.Wellknown(DTOCountry.Wellknowns.Spain)))
                            {
                                if (oZona.Provincia != null)
                                {
                                    string sProvincia = oZona.Provincia.Nom;
                                    if (sProvincia != "" & sProvincia != oLocation.Nom)
                                        sb.Append(string.Format(", {0}", sProvincia));
                                }
                            }
                            else
                            {
                                string sCountry = oCountry.LangNom.Esp;
                                if (!string.IsNullOrEmpty(sCountry))
                                {
                                    sb.Append(string.Format(", {0}", sCountry));
                                }

                            }
                        }
                    }
                }
            }

            var retval = sb.ToString();
            return retval;
        }

        public bool IsEmpty()
        {
            bool retval = this.Text == "" & this.ViaNom == "" & this.Num == "" & this.Zip == null;
            return retval;
        }

        public static List<string> Lines(DTOContact oContact)
        {
            List<string> retval = new List<string>();
            {
                retval.Add(string.IsNullOrEmpty(oContact.Nom) ? oContact.NomComercial : oContact.Nom);
                retval.Add(oContact.Address.Text);
                retval.Add(oContact.Address.Zip.ZipyCit());
            }
            return retval;
        }

        public static string ProvinciaOPais(DTOZip oZip)
        {
            string retval = "";
            if (oZip != null)
            {
                if (DTOArea.isEsp(oZip))
                {
                    List<Exception> exs = new List<Exception>();
                    var oProvincia = DTOZip.provincia(oZip);
                    if (oProvincia != null)
                        retval = "(" + oProvincia.Nom + ")";
                }
                else
                    retval = "(" + DTOArea.Country(oZip).LangNom.Esp + ")";
            }
            return retval;
        }

        public bool IsEsp()
        {
            bool retval = DTOArea.isEsp(this.Zip);
            return retval;
        }

        public static string GoogleText(DTOAddress oAddress)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            if (!string.IsNullOrEmpty(oAddress.ViaNom))
            {
                sb.Append(oAddress.ViaNom);
                sb.Append(",");
            }
            if (!string.IsNullOrEmpty(oAddress.Num))
            {
                sb.Append(oAddress.Num);
                sb.Append(",");
            }
            if (oAddress.Zip != null)
            {
                sb.Append(oAddress.Zip.ZipCod);
                sb.Append(",");
                sb.Append(oAddress.Location().Nom);
                sb.Append(",");
                sb.Append(oAddress.CountryNom(DTOLang.ESP()));
            }
            var retval = sb.ToString();
            return retval;
        }

        public static string GoogleNormalized(DTOAddress oSrc)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            if (oSrc != null && !string.IsNullOrEmpty(oSrc.Text))
            {
                string[] sAdrSegments = oSrc.Text.Split(' ');
                for (int i = 0; i < sAdrSegments.Count(); i++)
                {
                    // elimina el segon numero del carrer en cas de varios numeros

                    string segment = sAdrSegments[i];
                    if (VbUtilities.isNumeric(segment))
                    {
                        if (segment.Contains("-"))
                            segment = segment.Substring(0, segment.IndexOf("-"));
                    }
                }
                string sAdr = string.Join("+", sAdrSegments);
                string sCit = oSrc.Zip.Location.Nom.Replace(" ", "+");
                string sZona = oSrc.Zip.Location.Zona.Nom.Replace(" ", "+");
                string sCountry = oSrc.Zip.Location.Zona.Country.LangNom.Esp.Replace(" ", "+");
                sb.Append(sAdr);
                if (!string.IsNullOrEmpty(oSrc.Zip.ZipCod))
                    sb.Append("," + oSrc.Zip.ZipCod);
                sb.Append("," + sCit);
                sb.Append("," + sZona);
                sb.Append("," + sCountry);
            }
            string retval = sb.ToString();
            return retval;
        }

        public static string multiLine(DTOAddress oAddress)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            if (oAddress != null)
            {
                if (!string.IsNullOrEmpty(oAddress.Text))
                    sb.AppendLine(oAddress.Text.Trim());
                sb.Append(DTOZip.FullNom(oAddress.Zip));
            }
            string retval = sb.ToString();
            return retval;
        }

        public string ToHtml()
        {
            var sLines = multiLine(this);
            string retval = TextHelper.Html(sLines);
            return retval;
        }


        public static string ReverseSingleLine(DTOAddress oAddress, DTOLang oLang)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            string sLocation = DTOAddress.LocationFullNom(oAddress, oLang);
            if (!string.IsNullOrEmpty(sLocation))
            {
                sb.Append(sLocation);
                if (!string.IsNullOrEmpty(oAddress.Text))
                    sb.Append(", ");
            }
            sb.Append(oAddress.Text);
            string retval = sb.ToString();
            return retval;
        }

        public static string FullText(DTOAddress oAddress)
        {
            string retval = DTOAddress.multiLine(oAddress).toSingleLine();
            return retval;
        }

        public static string FullHtml(DTOAddress oAddress)
        {
            string retval = DTOAddress.multiLine(oAddress).toMultiLineHtml();
            return retval;
        }

        public bool Equals(DTOAddress candidate)
        {
            bool retval = false;
            if (candidate != null && this.Zip != null)
                retval = (this.Text == candidate.Text && this.Zip.Equals(candidate.Zip));
            return retval;
        }



        public DTOAddress Trimmed()
        {
            Type oType = this.GetType();
            var retval = Activator.CreateInstance(this.GetType());
            try
            {
                var oProperties = oType.GetProperties();
                foreach (var oProperty in oProperties)
                {
                    if (typeof(DTOBaseGuid).IsAssignableFrom(oProperty.PropertyType))
                    {
                        try
                        {
                            DTOBaseGuid oBaseGuid = (DTOBaseGuid)oProperty.GetValue(this);
                            if (oBaseGuid != null)
                            {
                                var oTrimmedPropertyValue = Activator.CreateInstance(oProperty.PropertyType, oBaseGuid.Guid);
                                oProperty.SetValue(retval, oTrimmedPropertyValue);
                            }
                        }
                        catch (Exception)
                        {
                            oProperty.SetValue(retval, oProperty.GetValue(this));
                        }
                    }
                    else
                        oProperty.SetValue(retval, oProperty.GetValue(this));
                }
            }
            catch (Exception)
            {
                System.Diagnostics.Debugger.Break();
            }
            return (DTOAddress)retval;
        }
    }
}
